### ISSUE TITLE
Correct way of using fp16 for training

### ISSUE THREAD

**Issue by bdalal** [allenai/longformer#33]
Hi,
Firstly, thank you for making the code and trained model public!

I have a question regarding the use of fp16 during training:
I'm using the longformer for sequence classification and I've created a pytorch module that adds a RobertaClassificationHead on top of the longformer like so:
```
class LongformerForSequenceClassification(Longformer):
    def __init__(self, pretrained_model, config):
        super(LongformerForSequenceClassification, self).__init__(config)
        self.num_labels = config.num_labels
        
        self.longformer = pretrained_model.half() # For fp16
        self.classifier = RobertaClassificationHead(config)
        
    def forward(self, input_ids=None, attention_mask=None, token_type_ids=None, position_ids=None, head_mask=None, input_embeds=None, labels=None,):
        outputs = self.longformer(input_ids, 
                                  attention_mask=attention_mask, 
                                  token_type_ids=token_type_ids, 
                                  position_ids=position_ids,
                                  head_mask=head_mask,)
        sequence_output = outputs[0]
        logits = self.classifier(sequence_output.float())
        
        outputs = (logits,) + outputs[2:]
        if labels is not None:
                loss_fct = CrossEntropyLoss()
                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
            outputs = (loss,) + outputs

        return outputs  # (loss), logits, (hidden_states), (attentions)
```
I'm halving the longformer model when I load it `self.longformer = pretrained_model.half() # For fp16`. This gets the code to work but my loss become `nan`. I'm guessing this is because of the numerical instabilities due to mixed precision training.

I've tried converting the entire model to fp16 as well the inputs but all of these fail at one or the other step because some aspect of the model (embedding, linear layers in classifier) do not permit this.

How can I correctly use fp16 in this context? From looking at the test for `sliding_chunks` should I just convert these 3 variables to fp16 https://github.com/allenai/longformer/blob/671f6e056dd2a3f51ac5843f8810f5ff68d5c85b/longformer/longformer.py#L127-L129 ?

I don't have Apex's AMP setup on my machine. I'm just using the functionality that pytorch offers.

Please let me know your thoughts. Thank you for your time!

**Comment by armancohan** [comment]
We encourage using apex for mixed precision training and not manually converting the model to half. With pytorch-lightning this is easy to do with only one additional `use_amp` argument passed to the `Trainer`. See the triviaqa script as an example: https://github.com/allenai/longformer/blob/master/scripts/triviaqa.py#L693

**Comment by bdalal** [comment]
Thanks for the quick response! For now, I have some systemic limitations because of which I'm unable to setup lightning and apex on my machine.
I did manage to halve the global and windowed linear layers at the line numbers I mentioned and that seems to work.
I'll evaluate using lighting as well once I get a chance.
Thanks again for your help!
